diff --git a/node_modules/cypress/lib/tasks/verify.js b/node_modules/cypress/lib/tasks/verify.js
index 7dfbe4c..887c704 100644
--- a/node_modules/cypress/lib/tasks/verify.js
+++ b/node_modules/cypress/lib/tasks/verify.js
@@ -323,7 +323,7 @@ var isRootUser = function isRootUser() {
  * @see https://crbug.com/638180
 */
 var needsSandbox = function needsSandbox() {
-  return isLinuxLike() && isRootUser();
+  return true;
 };
 
 module.exports = {
